CONTRIBUTING.md

```diff
@@ -4,27 +4,63 @@ To get up and running, install the dependencies and run the tests:
+Each test directory in `tests` has a `jsfmt.spec.js` file that controls how exactly the rest of the files in the directory are used for tests. This file must contain one or more calls to the `run_spec` global function. For example, in directories with JavaScript formatting tests, `jsfmt.spec.js` generally looks like this:
 
-If you can, take look at [commands.md](commands.md) and check out [Wadler's paper](http://homepages.inf.ed.ac.uk/wadler/papers/prettier/prettier.pdf) to understand how Prettier works.
+```js
+run_spec(__dirname, ["babel", "flow", "typescript"]);`
+```
+
+This verifies that for each file in the directory, the output matches the snapshot and is the same for each listed parser.
+
+You can also pass options as the third argument:
+
+```js
+run_spec(__dirname, ["babel"], { trailingComma: "es5" });
+```
+
+Signature:
+
+```ts
+function run_spec(
+  fixtures: { dirname: string; snippets?: string[] } | string,
+  parsers: string[],
+  options?: PrettierOptions & {
+    errors: true | { [parserName: string]: true | string[] };
+  }
+): void;
+```
+
+Parameters:
+
+- `fixtures`: Must be set to `__dirname` or to an object of the shape `{ dirname: __dirname, ... }`. The object may have the `snippets` field to specify extra input entries as a string array in addition to the files in the current directory. For each input entry (a file or a snippet) `run_spec` configures and runs a number of tests. The main check is that for a given input the output should match the snapshot. [Additional checks](#deeper-testing) are controlled by options and environment variables.
+- `parsers`: A list of parser names. The tests verify that the parsers in this list produce the same output. If the list includes `typescript`, then `babel-ts` is included implicitly. If the list includes `babel`, and the current directory is inside `tests/js`, then `espree` and `meriyah` are included implicitly.
+- `options`: In addition to Prettier's formatting option, can contain the `errors` property to specify that it's expected that the formatting shouldn't be successful and an error should be thrown for all (`entries: true`) or some combinations of input entries and parsers.
+
+The implementation of `run_spec` can be found in `tests_config/run_spec.js`.
+
+`tests/flow-repo/` contains the Flow test suite and is not supposed to be edited by hand. To update it, clone the Flow repo next to the Prettier repo and run: `node scripts/sync-flow-tests.js ../flow/tests/`.
+
+## Debugging
+
+To debug Prettier locally, you can either debug it in Node (recommended) or the browser.
+
+- The easiest way to debug it in Node is to create a local test file with some example code you want formatted and either run it in an editor like VS Code or run it directly via `./bin/prettier.js <your_test_file>`.
+- The easiest way to debug it in the browser is to build Prettier's website locally (see [`website/README.md`](website/README.md)).
+
+## Other
+
+Run `yarn fix` to lint and format all files.
```

Comment:
I think  better suggest use `yarn lint`, if there is bug  in those eslint rules,        they may lost their work.

Reply:
If you're not sure about some rules, let's better disable them. Many people (me included) use autofix on save. I don't want to lose work too.

Reply:
I think enabled rules don't have known issues, in editor if you found something not right you can undo, but if editor already closed and haven't committed, run `yarn fix` you can't find your original code back.

Reply:
I just might not notice the editor did that.